[WIP] Variable length trials fix

sliwy opened the pull request:
Just two small changes to fix cropped decoding to work with variable length trials. Tests does not work locally in my case (no clue why), so let's check here. Trying to solve #222

robintibor commented on braindecode/classifier.py:
```diff
@@ -221,7 +221,7 @@ def predict_proba(self, X):
         """
         y_pred = super().predict_proba(X)
-        if self.cropped:
+        if self.cropped and len(y_pred.shape) == 3:
```
Please add some comment here why this happens! And what we expect etc....

robintibor replied:
```suggestion
        if self.cropped and len(y_pred.shape) == 3:
```

```suggestion
# Normally, we have to average the predictions across crops/timesteps
# to get one prediction per time window/trial
# In the cropped scoring , we may have averaged these predictions 
# already, hence the check if len(y_pred.shape) == 3:
        if self.cropped and len(y_pred.shape) == 3:
```

sliwy replied:
this is what I left for later (together with the comments to do) if the changes from this PR works with your example. @robintibor have you checked if ti works with your example with variable trial length?

robintibor replied:
I checked it in my mind and it seemed to work! Now also checked in reality and also worked! :)

robintibor replied:
so do you have capacity to finish it? @sliwy  Otherwise I can have a try :)

sliwy replied:
@robintibor I can take care of it today or tomorrow evening. If you need it faster feel free to finish that. :)

robintibor replied:
sounds good! should be quick enough! :)

robintibor commented on braindecode/training/scoring.py:
```diff
@@ -177,7 +177,9 @@ def on_epoch_end(self, net, dataset_train, dataset_valid, **kwargs):
             # trial preds is a list
             # each item is an 2d array classes x time
-            y_preds_per_trial = np.array(trial_preds)
+            y_preds_per_trial = np.array(
```
Also here some comments why we are averaging here would be great I think

robintibor commented:
Great I think this can be merged, but would be great if you can add some comments so we know what we were doing there and why

robintibor commented on braindecode/training/scoring.py:
```diff
@@ -177,7 +177,9 @@ def on_epoch_end(self, net, dataset_train, dataset_valid, **kwargs):
             # trial preds is a list
             # each item is an 2d array classes x time
-            y_preds_per_trial = np.array(trial_preds)
+            y_preds_per_trial = np.array(
+                [np.mean(p, axis=1) for p in trial_preds]
+            )
```
```suggestion
            # Average across the timesteps of each trial so we have per-trial predictions already,
            # these will be just passed through the forward method of the classifier/regressor
            # to the skorch scoring function
            y_preds_per_trial = np.array(
                [np.mean(p, axis=1) for p in trial_preds]
            )
```

robintibor replied:
Do we also need to adapt the forward function of the regressor @sliwy ?

sliwy commented:
I added comments you asked for, I also modified EEGRegressor to work the same way as EEGClassifier. 

I still have some doubts whether the `predict_proba` should work with variable length trials (for scoring it should be ok). In particular, I don't know what will be the output of skorch `predict_proba` method when provided with observation of different shape. I've never done this before. Maybe we should create a test for that? It would need a custom DataLoader?  @robintibor  As you've already tested it on your side maybe you have a minimal working example for variable length trials decoding that we can use a starting point for the test?

robintibor commented:
I think my example is too complex, also requires loading TUH etc., but I can also see whether I can distill something out of it, if you have time before to make a test, also great, otherwise I will see

robintibor commented:
ok since my colleague needs it quite urgently now, merging this and we can do test in separate PR potentially.
